webpack/scenes/ContentViews/Publish/PublishContentViewWizard.js

```diff
@@ -0,0 +1,134 @@
+import CVPublishReview from './CVPublishReview';
+import {
+  selectEnvironmentPaths,
+  selectEnvironmentPathsStatus,
+} from '../components/EnvironmentPaths/EnvironmentPathSelectors';
+import './publishWizardHeader.scss';
+
+const PublishContentViewWizard = ({
+  details, show, setIsOpen, currentStep, setCurrentStep,
+}) => {
+  const { name, id: cvId, version_count: versionCount } = details;
+  const [description, setDescription] = useState('');
+  const [checkedItems, setCheckedItems] = useState([]);
+  const [promote, setPromote] = useState(false);
+  const [forcePromote, setForcePromote] = useState([]);
+  const [envPathFlat, setEnvPathFlat] = useState([]);
+  const dispatch = useDispatch();
+  const environmentPathResponse = useSelector(selectEnvironmentPaths);
+  const environmentPathStatus = useSelector(selectEnvironmentPathsStatus);
+  const environmentPathLoading = environmentPathStatus === STATUS.PENDING;
+
+  const steps = [
+    {
+      id: 1,
+      name: 'Publish',
+      component: <CVPublishForm
+        description={description}
+        setDescription={setDescription}
+        details={details}
+        checkedItems={checkedItems}
+        setCheckedItems={setCheckedItems}
+        promote={promote}
+        setPromote={setPromote}
+        forcePromote={forcePromote}
+      />,
+    },
+    {
+      id: 2, name: 'Review', component: <CVPublishReview details={details} checkedItems={checkedItems} />, nextButtonText: 'Finish',
+    },
+    {
+      id: 3,
+      name: 'Finish',
+      component: <CVPublishFinish
+        description={description}
+        setDescription={setDescription}
+        checkedItems={checkedItems}
+        setCheckedItems={setCheckedItems}
+        forcePromote={forcePromote}
+        cvId={cvId}
+        versionCount={versionCount}
+        show={show}
+        setIsOpen={setIsOpen}
+        currentStep={currentStep}
+        setCurrentStep={setCurrentStep}
+      />,
+      isFinishedStep: true,
+    },
+  ];
+
+  useEffect(
+    () => {
+      dispatch(getEnvironmentPaths());
+      setCurrentStep(1);
+    },
+    [],
+  );
+
+  useEffect(() => {
+    if (!environmentPathLoading) {
+      const { results } = environmentPathResponse || {};
+      setEnvPathFlat(results.map(result => result.environments).flatten());
+    }
+  }, [environmentPathResponse, environmentPathLoading]);
+
+  const getEnv = envId => envPathFlat.find(item => item.id === envId);
+
+  const isValid = (env) => {
+    if (!env.prior || env.prior.name === 'Library') return true;
+    if (!checkedItems.includes(getEnv(env.prior.id)) && !(env.prior.name === 'Library')) return false;
+    return checkedItems.includes(env) && isValid(getEnv(env.prior.id));
+  };
+
+  useEffect(() => {
+    const forcedList = [];
+    checkedItems.forEach((item) => {
+      if (!isValid(item)) forcedList.push(item);
+    });
+    setForcePromote(forcedList);
+  }, [checkedItems, setForcePromote]);
```

Comment:
```
useEffect(() => {
    setForcePromote(checkedItems.map((item) => {
      if (!isValid(item)) return item;
    }).filter(item => item));
  }, [checkedItems, setForcePromote]);
```

This works too. Let me know which is better..

Reply:
I like a combination of the two:
1. Use the `forcedList` variable name to explain what we're doing;
2. Use `map` to consolidate and avoid using `forEach` + `push`.  _Update:_ I think it can be simplified even more by just using `filter`:
```suggestion
  useEffect(() => { 
   const forcedList = checkedItems.filter(item => !isValid(item)); 
   
   setForcePromote(forcedList);
   }, [checkedItems, setForcePromote]);
```